### PR TITLE
Support mrope triton kernel and add unit test

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

Introduce triton mrope kernel in rotary embedding. 
Before:
image

After:
image

Adapted from vLLM.

```
Unit test passed:

$python -m pytest test_mrope.py
============================================================================================= test session starts =============================================================================================
platform linux -- Python 3.10.13, pytest-8.3.5, pluggy-1.5.0
rootdir: /root/luoyuan.luo/workspace/sglang_dev/test/srt/rotary_embedding
plugins: anyio-4.8.0
collected 8 items                                                                                                                                                                                             

test_mrope.py ........                                                                                                                                                                                  [100%]

============================================================================================= 8 passed in 13.60s ==============================================================================================
gsm8k acc no drop

$python3 benchmark/gsm8k/bench_sglang.py --num-questions 200 --parallel 128 --num-shots 8 --port 30000 --data-path ./test.jsonl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:14<00:00, 13.73it/s]
Accuracy: 0.950
Invalid: 0.000
Latency: 14.814 s
Triton kernel vs baseline Torch baseline：
TP8 speedup 32%
TP4 speedup 45%
TP2 speedup 48%
More benchmark data will be updated.

TP8
$python3 benchmark_mrope.py --model-name /home/admin/Qwen2.5-VL-72B-Instruct --tp-size 8 --warmup-iter 10 --benchmark-iter 100 --dtype bfloat16 --seed 0 --num-tokens 1024
Namespace(model_name='/home/admin/Qwen2.5-VL-72B-Instruct', tp_size=8, warmup_iter=10, benchmark_iter=100, dtype='bfloat16', seed=0, num_tokens=[1024], trust_remote_code=False, output_csv='mrope_benchmark_results.csv')
================================================================================
Evaluating model: /home/admin/Qwen2.5-VL-72B-Instruct with tp_size: 8 and num_tokens: 1024, dtype: torch.bfloat16

Performance for config (1024, 8, 1):
Torch implementation: mean=0.00016325s, median=0.00016177s, p99=0.00018089s
Triton implementation: mean=0.00012344s, median=0.00012112s, p99=0.00014963s
Triton Speedup over Torch: 1.32250401x
Benchmark results saved to mrope_benchmark_results_20251010_192016.csv

TP4
$python3 benchmark_mrope.py --model-name /home/admin/Qwen2.5-VL-72B-Instruct --tp-size 4 --warmup-iter 10 --benchmark-iter 100 --dtype bfloat16 --seed 0 --num-tokens 1024
Namespace(model_name='/home/admin/Qwen2.5-VL-72B-Instruct', tp_size=4, warmup_iter=10, benchmark_iter=100, dtype='bfloat16', seed=0, num_tokens=[1024], trust_remote_code=False, output_csv='mrope_benchmark_results.csv')
================================================================================
Evaluating model: /home/admin/Qwen2.5-VL-72B-Instruct with tp_size: 4 and num_tokens: 1024, dtype: torch.bfloat16

Performance for config (1024, 16, 2):
Torch implementation: mean=0.00017449s, median=0.00017321s, p99=0.00020296s
Triton implementation: mean=0.00011901s, median=0.00011730s, p99=0.00014256s
Triton Speedup over Torch: 1.46622326x
Benchmark results saved to mrope_benchmark_results_20251010_193246.csv

TP2
$python3 benchmark_mrope.py --model-name /home/admin/Qwen2.5-VL-72B-Instruct --tp-size 2 --warmup-iter 10 --benchmark-iter 100 --dtype bfloat16 --seed 0 --num-tokens 1024
Namespace(model_name='/home/admin/Qwen2.5-VL-72B-Instruct', tp_size=2, warmup_iter=10, benchmark_iter=100, dtype='bfloat16', seed=0, num_tokens=[1024], trust_remote_code=False, output_csv='mrope_benchmark_results.csv')
================================================================================
Evaluating model: /home/admin/Qwen2.5-VL-72B-Instruct with tp_size: 2 and num_tokens: 1024, dtype: torch.bfloat16

Performance for config (1024, 32, 4):
Torch implementation: mean=0.00018627s, median=0.00018525s, p99=0.00019919s
Triton implementation: mean=0.00012546s, median=0.00012255s, p99=0.00016031s
Triton Speedup over Torch: 1.48466421x
Benchmark results saved to mrope_benchmark_results_20251010_193415.csv
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by yuan-luo** [comment]
I have a PR referenced by vllm without adding any reference information in code segment. So this time I decide not to add reference as well.

**Comment by b8zhong** [comment]
Yuan, I think https://github.com/sgl-project/sglang/pull/11031 achieves the same objective, as I think the vLLM one was also from Liger (But feel free to merge yours if it can be merged faster than the one I opened)

**Comment by yuan-luo** [comment]
> Yuan, I think #11031 achieves the same objective, as I think the vLLM one was also from Liger (But feel free to merge yours if it can be merged faster than the one I opened)

@b8zhong Maybe we can adopt this PR, as it adds benchmark and accuracy test. The perf we can gain in small model are close, I'll add you as co-author.

**Comment by b8zhong** [comment]
@yuan-luo Sure, sounds good

**Comment by yuan-luo** [comment]
> @yuan-luo Sure, sounds good

Done.

File: python/sglang/srt/layers/rotary_embedding.py [hunk 1]
```diff
@@ -1141,6 +1328,51 @@ def forward(
...
+    def forward(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        fused_set_kv_buffer_arg: Optional[FusedSetKVBufferArg] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        assert positions.ndim == 1 or positions.ndim == 2
+        assert key is not None
+
```
[the remark this answers is not in the document]
Comment (BBuf) [reply python/sglang/srt/layers/rotary_embedding.py:1346]: +1

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1033,6 +1035,182 @@ def apply_interleaved_rope(x: torch.Tensor, mrope_section: list[int]) -> torch.T
...
+        k: [num_tokens, num_kv_heads * head_size]
+        cos: [3, num_tokens, head_size //2 ]
+            (T/H/W positions with multimodal inputs)
+        sin: [3, num_tokens, head_size //2 ]
+            (T/H/W positions with multimodal inputs)
+        mrope_section: [t, h, w]
+        head_size: int
+    """
+    n_row, n_q_head_head_dim = q.shape
+    n_q_head = n_q_head_head_dim // head_size
```
Comment (BBuf) [inline python/sglang/srt/layers/rotary_embedding.py:1183]: Maybe add check is better, such as:

```python
def triton_mrope(...):
    n_row, n_q_head_head_dim = q.shape
    assert n_q_head_head_dim % head_size == 0, \
        f"q shape {n_q_head_head_dim} must be divisible by head_size {head_size}"
    n_q_head = n_q_head_head_dim // head_size
    
    assert k.shape[1] % head_size == 0, \
        f"k shape {k.shape[1]} must be divisible by head_size {head_size}"
    n_kv_head = k.shape[1] // head_size
```

File: python/sglang/srt/layers/rotary_embedding.py [hunk 3]
```diff
@@ -1033,6 +1035,182 @@ def apply_interleaved_rope(x: torch.Tensor, mrope_section: list[int]) -> torch.T
...
+    q: torch.Tensor,
+    k: torch.Tensor,
+    cos: torch.Tensor,
+    sin: torch.Tensor,
+    mrope_section: list[int],
+    head_size: int,
+    rotary_dim: int,
+    mrope_interleaved: bool,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Qwen2VL mrope kernel.
```
Comment (BBuf) [inline python/sglang/srt/layers/rotary_embedding.py]: Only used in Qwen2VL?

File: benchmark/kernels/rotary_embedding/benchmark_mrope.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        for tp_size in tp_list:
+            config = get_model_config(model_name)
+            # get the model config
+            total_num_kv_heads = config.num_key_value_heads
+            total_num_heads = config.num_attention_heads
+            num_heads = total_num_heads // tp_size
+            num_kv_heads = max(1, total_num_kv_heads // tp_size)
+            head_dim = config.hidden_size // total_num_heads
+            q_size = num_heads * head_dim
+            kv_size = num_kv_heads * head_dim
```
[the remark this answers is not in the document]
Comment (BBuf) [reply benchmark/kernels/rotary_embedding/benchmark_mrope.py]: +1

File: python/sglang/srt/layers/rotary_embedding.py [hunk 5]
```diff
@@ -1141,6 +1328,51 @@ def forward(
...
+        query_pass = query[..., self.rotary_dim :]
+        query_rot = _apply_rotary_emb(query_rot, cos, sin, self.is_neox_style)
+        query = torch.cat((query_rot, query_pass), dim=-1).reshape(query_shape)
+
+        key = key.view(num_tokens, -1, self.head_size)
+        key_rot = key[..., : self.rotary_dim]
+        key_pass = key[..., self.rotary_dim :]
+        key_rot = _apply_rotary_emb(key_rot, cos, sin, self.is_neox_style)
+        key = torch.cat((key_rot, key_pass), dim=-1).reshape(key_shape)
+        return query, key
```
[the remark this answers is not in the document]
Comment (BBuf) [reply python/sglang/srt/layers/rotary_embedding.py:1380]: Make sense.

File: sgl-kernel/benchmark/bench_mrope.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
+# Adapted from vLLM benchmark_mrope.py
```
Comment (BBuf) [inline sgl-kernel/benchmark/bench_mrope.py:1]: Remove it to sgl-kernel benchmark and get ci runner to avoid dead benchmark code.

**Review by BBuf — APPROVED** [review]
LGTM, fix the comments is better.

**Review by Alcanderian — APPROVED** [review]
